src/main/scala/z3/scala/dsl/ValHandler.scala

```diff
@@ -20,9 +20,9 @@ abstract class ValHandler[A : Default] {
   /** Z3 code to construct a sort representing the Scala A type. */
   def mkSort(z3 : Z3Context) : Z3Sort
 
-  /** Constructs a Val[A], i.e. the representation of a variable of type A. */
-  def construct : Val[A] = new Val[A] {
-    def build(z3 : Z3Context) = z3.mkFreshConst("valCst", mkSort(z3))
+  /** Constructs a ValTree[A], i.e. the representation of a variable of type A. */
+  def construct[T >: BottomSort <: TopSort]: Tree[T] = new ValTree[T] {
```

Comment:
This signature seems a bit weird. Doesn't this mean we can construct `ValTree` instances with different sorts from a same `ValHandler`? Maybe we should introduce some implicit evidence that binds `Default` instances to sorts. This can be done by the `ValHandler` if you want, by giving it a second type parameter for example.